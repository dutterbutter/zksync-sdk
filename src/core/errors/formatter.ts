// src/core/errors/formatter.ts

/* -------------------- Formatting helpers -------------------- */
import type { ErrorEnvelope } from '../types';
import chalk from 'chalk';

function elideMiddle(s: string, max = 96): string {
  if (s.length <= max) return s;
  const keep = Math.max(10, Math.floor((max - 1) / 2));
  return `${s.slice(0, keep)}…${s.slice(-keep)}`;
}

function shortJSON(v: unknown, max = 240): string {
  try {
    const s = JSON.stringify(v, (_k: string, val: unknown): unknown =>
      typeof val === 'bigint' ? `${val.toString()}n` : val,
    );
    return s.length > max ? elideMiddle(s, max) : s;
  } catch {
    return String(v);
  }
}

function kv(label: string, value: string): string {
  const width = 10;
  const pad = label.length >= width ? ' ' : ' '.repeat(width - label.length);
  return `${chalk.dim(label + pad)}: ${value}`;
}

function formatContextLine(ctx?: Record<string, unknown>): string | undefined {
  if (!ctx) return;
  const txHash = ctx['txHash'] ?? ctx['l1TxHash'] ?? ctx['hash'];
  const nonce = ctx['nonce'];
  const parts: string[] = [];
   
  if (txHash !== undefined)
    parts.push(`txHash=${typeof txHash === 'string' ? txHash : shortJSON(txHash, 96)}`);
  if (nonce !== undefined) {
    const nonceStr =
      typeof nonce === 'string' || typeof nonce === 'number' || typeof nonce === 'bigint'
        ? String(nonce)
        : shortJSON(nonce, 48);
    parts.push(`nonce=${nonceStr}`);
  }
  return parts.length ? `  ${kv('Context', parts.join('  •  '))}` : undefined;
}

function formatStep(ctx?: Record<string, unknown>): string | undefined {
  const step = ctx && typeof ctx['step'] === 'string' ? ctx['step'] : undefined;
  return step ? `  ${kv('Step', step)}` : undefined;
}

function formatRevert(r?: ErrorEnvelope['revert']): string | undefined {
  if (!r) return;
  const first = [`selector=${r.selector}`];
  const lines: string[] = [];
  lines.push(`  ${kv('Revert', first.join(' '))}`);
  if (r.name) lines.push(`              ${chalk.cyan('name=')}${r.name}`);
  if (r.contract) lines.push(`              ${chalk.cyan('contract=')}${r.contract}`);
  if (r.fn) lines.push(`              ${chalk.cyan('fn=')}${r.fn}`);
  if (r.args && r.args.length) {
    lines.push(`              ${chalk.cyan('args=')}${shortJSON(r.args, 120)}`);
  }
  return lines.join('\n');
}

function formatCause(c?: unknown): string[] {
  if (!c) return [];
  const out: string[] = [];

  // If the cause is an object, read known fields safely; otherwise stringify it.
  if (typeof c === 'object' && c !== null) {
    const obj = c as Record<string, unknown>;
    const head: string[] = [];
    if (obj.name !== undefined) {
      const nameVal = obj.name;
      const nameStr =
        typeof nameVal === 'string' ||
        typeof nameVal === 'number' ||
        typeof nameVal === 'bigint' ||
        typeof nameVal === 'boolean'
          ? String(nameVal)
          : shortJSON(nameVal, 120);
      head.push(`name=${nameStr}`);
    }
    if (obj.code !== undefined) {
      const codeVal = obj.code;
      const codeStr =
        typeof codeVal === 'string' ||
        typeof codeVal === 'number' ||
        typeof codeVal === 'bigint' ||
        typeof codeVal === 'boolean'
          ? String(codeVal)
          : shortJSON(codeVal, 120);
      head.push(`code=${codeStr}`);
    }
    if (head.length) out.push(`  ${kv('Cause', head.join('  '))}`);

    if (obj.message) {
      const messageStr =
        typeof obj.message === 'string' ||
        typeof obj.message === 'number' ||
        typeof obj.message === 'bigint' ||
        typeof obj.message === 'boolean'
          ? String(obj.message)
          : shortJSON(obj.message, 600);
      out.push(`              message=${elideMiddle(messageStr, 600)}`);
    }
    if (obj.data) {
      const dataStr = shortJSON(obj.data, 200);
      out.push(`              data=${elideMiddle(dataStr, 200)}`);
    }
  } else {
    out.push(`  ${kv('Cause', shortJSON(c, 200))}`);
  }

  return out;
}

export function formatEnvelopePretty(e: ErrorEnvelope): string {
  const lines: string[] = [];

  // Header
  lines.push(`${chalk.red('✖')} ${chalk.bold('ZKsyncError')} [${chalk.yellow(e.type)}]`);
  lines.push(`  ${kv('Message', e.message)}`);
  lines.push('');

  lines.push(`  ${kv('Operation', e.operation)}`);

  const ctx = (() => {
    const u = e as unknown;
    if (!u || typeof u !== 'object') return undefined;
    const obj = u as Record<string, unknown>;
    const candidate = obj['ctx'] ?? obj['context'];
    if (candidate && typeof candidate === 'object' && candidate !== null) {
      return candidate as Record<string, unknown>;
    }
    return undefined;
  })();

  const ctxLine = formatContextLine(ctx);
  if (ctxLine) lines.push(ctxLine);

  const stepLine = formatStep(ctx);
  if (stepLine) lines.push(stepLine);

  const rv = formatRevert(e.revert);
  if (rv) lines.push(rv);

  const causeLines = formatCause(e.cause);
  if (causeLines.length) {
    if (!ctxLine && !rv) lines.push('');
    lines.push(...causeLines);
  }

  return lines.join('\n');
}
