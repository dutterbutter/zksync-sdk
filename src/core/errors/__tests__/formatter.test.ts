/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
import { describe, it, expect, beforeAll } from 'bun:test';
import { formatEnvelopePretty } from '../formatter';
import chalk from 'chalk';

beforeAll(() => {
  // Stabilize snapshots by disabling colors
  chalk.level = 0;
});

describe('errors/formatter.formatEnvelopePretty', () => {
  it('formats a rich envelope with context, step, revert, and cause', () => {
    const envelope = {
      type: 'RPC',
      message: 'Failed to fetch L2â†’L1 log proof.',
      operation: 'zksrpc.getL2ToL1LogProof',
      resource: 'zksrpc',
      context: {
        txHash: '0x' + 'aa'.repeat(32),
        nonce: 7,
        step: 'fetch-proof',
      },
      revert: {
        selector: '0x08c379a0',
        name: 'Error',
        contract: 'L2MessageVerification',
        fn: 'verify(bytes32)',
        args: ['0x' + '11'.repeat(32)],
      },
      cause: {
        name: 'TimeoutError',
        code: 'ETIMEDOUT',
        message: '5000ms exceeded while waiting for response',
        data: '0xdeadbeefcafebabedeadbeef', // long, will be elided
      },
    } as const;

    const pretty = formatEnvelopePretty(envelope as any);
    expect(pretty).toMatchSnapshot();
  });

  it('handles minimal envelope without optional fields', () => {
    const envelope = {
      type: 'STATE',
      message: 'Proof not yet available. Please try again later.',
      operation: 'zksrpc.getL2ToL1LogProof',
      resource: 'zksrpc',
    };

    const pretty = formatEnvelopePretty(envelope as any);
    expect(pretty).toMatchSnapshot();
  });
});
