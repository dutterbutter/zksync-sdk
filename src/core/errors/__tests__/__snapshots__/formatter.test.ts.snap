// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`errors/formatter.formatEnvelopePretty formats a rich envelope with context, step, revert, and cause 1`] = `
"✖ ZKsyncError [RPC]
  Message   : Failed to fetch L2→L1 log proof.

  Operation : zksrpc.getL2ToL1LogProof
  Resource  : zksrpc
  Step      : fetch-proof

  Context   : txHash=0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa  •  nonce=7
  Revert    : selector=0x08c379a0
              name=Error
              contract=L2MessageVerification
              fn=verify(bytes32)
              args=["0x1111111111111111111111111111111111111111111111111111111111111111"]
  Cause     : name=TimeoutError  code=ETIMEDOUT
              message=5000ms exceeded while waiting for response
              data=0xdeadbeefcafebabedeadbeef"
`;

exports[`errors/formatter.formatEnvelopePretty handles minimal envelope without optional fields 1`] = `
"✖ ZKsyncError [STATE]
  Message   : Proof not yet available. Please try again later.

  Operation : zksrpc.getL2ToL1LogProof
  Resource  : zksrpc
"
`;
